0.2.0/docker-entrypoint.sh

```diff
@@ -251,26 +40,32 @@ if [ "$1" == 'janusgraph' ] && [ "$(id -u)" == "0" ]; then
     exec chroot --skip-chdir --userspec janusgraph:janusgraph / "${BASH_SOURCE}" "$@"
 fi
 
-# https://gist.github.com/samizuh/a978bc530e88847c2b5d8f75cb8caee0
-
 # running as non root user
 if [ "$1" == 'janusgraph' ]; then
+    # setup config directory
+    mkdir -p ${JANUS_CONFIG_DIR}
+    cp ${JANUS_HOME}/conf/gremlin-server/janusgraph-${JANUS_PROPS_TEMPLATE}-server.properties ${JANUS_CONFIG_DIR}/janusgraph.properties
+    cp ${JANUS_HOME}/conf/gremlin-server/gremlin-server.yaml ${JANUS_CONFIG_DIR}
 
     mkdir -p ${JANUS_DATA_DIR}
     chown -R "$(id -u):$(id -g)" ${JANUS_DATA_DIR} ${JANUS_CONFIG_DIR}
     chmod 700 ${JANUS_DATA_DIR} ${JANUS_CONFIG_DIR}
     chmod -R 600 ${JANUS_CONFIG_DIR}/*
 
-    writer_server_props
-    write_storage_props
-
-    ORIG_IFS=${IFS}
-    IFS=', ' read -r -a index_names <<< "${JANUS_INDEXES}"
-    IFS=${ORIG_IFS}
+    # apply configuration from environment
+    while IFS='=' read -r envvar_key envvar_val; do
+      if [[ "$envvar_key" =~ server\. ]] && [[ ! -z $envvar_val ]]; then
+        write_prop $GREMLIN_YAML
+      elif [[ "$envvar_key" =~ ^[a-z0-9_]+\.[a-z0-9_]+ ]]; then
+        write_prop $JANUS_PROPS
+      fi
+    done < <(env)
 
-    for idx in ${index_names}
-    do
-      write_index_props ${idx}
+    # wait for graph
```

Comment:
The server doesn't connect properly and it doesn't appear to recover if the backend isn't available at startup. Interestingly the server does still startup.

```
jce-janusgraph   | 1263 [main] WARN  org.apache.tinkerpop.gremlin.server.GremlinServer  - Graph [graph] configured at [/etc/opt/janusgraph/janusgraph.properties] could not be instantiated and will not be available in Gremlin Server.  GraphFactory message: GraphFactory could not instantiate this Graph implementation [class org.janusgraph.core.JanusGraphFactory]
jce-janusgraph   | java.lang.RuntimeException: GraphFactory could not instantiate this Graph implementation [class org.janusgraph.core.JanusGraphFactory]
...
jce-janusgraph   | Caused by: com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: jce-cassandra/192.168.208.2:9042 (com.datastax.driver.core.exceptions.TransportException: [jce-cassandra/192.168.208.2:9042] Cannot connect))
...
jce-janusgraph   | 4082 [gremlin-server-boss-1] INFO  org.apache.tinkerpop.gremlin.server.GremlinServer  - Channel started at port 8182.
```

So I think it makes sense to wait for the backend to be available before trying to start the server, as that's not in the clients control to correct. But I agree it should also properly handle the case where the issue is the user's config. How about logging the errors during connection attempts while waiting and adding a timeout setting?

Reply:
> The server doesn't connect properly and it doesn't appear to recover if the backend isn't available at startup. 

But should it recover if the backend isn't available? I think it should simply fail in this case. This gives users quick feedback if they start the container manually with `docker run` and in production environments, the container will likely be started again until it eventually works when the backend is back online.
I don't have a strong opinion on this one though. So, if you have a strong feeling on this or good arguments for why the container should wait for the backend, then we can also keep it like this.

> Interestingly the server does still startup.

I also didn't expect that to be honest. The reasoning is probably that users can configure multiple graphs and expect the server to make all available that didn't produce any errors.
That probably also means that we should think about a better healthcheck that not only relies on the availability of port `8182`, at least for a future release.

Reply:
I'll think some more on possible better approaches here. I can see the case for failing if the storage/indexing backends are external. I'm thinking about the case where you're managing the storage/indexing backend as containers together with your JaunsGraph instance, e.g. [docker-compose-cql-es.yaml](https://github.com/sjudeng/janusgraph-docker/blob/starter-config-updates/docker-compose-cql-es.yml). For that case if you remove the wait-for-graph logic you can see it always fails with the above error and so you'd always need to restart the janusgraph service manually to get it functioning.

Reply:
OK, Docker Compose is a good example where the approach to wait makes sense. That even seems to be the recommended approach in [the official Docker docs](https://docs.docker.com/compose/startup-order/):

> To handle this, design your application to attempt to re-establish a connection to the database after a failure. If the application retries the connection, it can eventually connect to the database.

> The best solution is to perform this check in your application code, both at startup and whenever a connection is lost for any reason. However, if you don’t need this level of resilience, you can work around the problem with a wrapper script:

They even show a bash script that is very similar to what you have here.

So, your suggestion from earlier is probably really the best option:

> How about logging the errors during connection attempts while waiting and adding a timeout setting?

Reply:
I added support for a `JANUS_STORAGE_TIMEOUT` environment variable that if set will do the wait-for-graph logic with the given timeout. Currently by default it's unset so it won't test the connection and will just immediately start the server. It does mean that users are responsible for setting it if needed. Personally I'd prefer instead to configure the default to 60s, as then it'll work for the above Compose-type cases, which might be more typical for new users, but also still attempt to start the server (and log the result/errors) after the timeout. Users could still unset it if it doesn't make sense for their environment. Do you have a preference?

Reply:
> Personally I'd prefer instead to configure the default to 60s

I'm fine with that